File: pisa-proxy/http/src/http.rs
// Copyright 2022 SphereEx Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::net::{IpAddr, Ipv4Addr, SocketAddr};

use axum::{
    body::Body,
    extract::State,
    http::{header, StatusCode},
    response::Response,
    routing::get,
    Router,
};
use config::config::PisaProxyConfig;
use pisa_error::error::*;
use pisa_metrics::metrics::MetricsManager;
use tracing::info;

#[async_trait::async_trait]
pub trait HttpServer {
    async fn start(&mut self) -> Result<(), Error>;
}

pub trait HttpFactory {
    fn build_http_server(&self, kind: HttpServerKind) -> impl HttpServer;
}

pub enum HttpServerKind {
    Axum,
}

pub struct PisaHttpServerFactory {
    pisa_config: PisaProxyConfig,
    metrics_manager: MetricsManager,
}
impl PisaHttpServerFactory {
    pub fn new(pcfg: PisaProxyConfig, mgr: MetricsManager) -> PisaHttpServerFactory {
        PisaHttpServerFactory { pisa_config: pcfg, metrics_manager: mgr }
    }
}

impl HttpFactory for PisaHttpServerFactory {
    fn build_http_server(&self, kind: HttpServerKind) -> impl HttpServer {
        match kind {
            HttpServerKind::Axum => AxumServer {
                pisa_config: self.pisa_config.clone(),
                metrics_manager: self.metrics_manager.clone(),
            },
        }
    }
}

pub async fn new_http_server(mut s: impl HttpServer) {
    s.start().await.unwrap();
}

#[derive(Clone, Debug)]
pub struct AxumServer {
    pisa_config: PisaProxyConfig,
    metrics_manager: MetricsManager,
}

impl AxumServer {
    fn routes(&self) -> Router<(), Body> {
        let state = self.clone();

        Router::new()
            .route("/healthz", get(Self::healthz))
            .route("/version", get(Self::version))
            .route("/metrics", get(Self::metrics))
            .with_state(state)
    }

    async fn healthz(_state: State<Self>) -> StatusCode {
        // TODO: add checking logic
        StatusCode::OK
    }

    async fn version(State(state): State<Self>) -> String {
        state.pisa_config.get_version().to_string()
    }

    async fn metrics(State(state): State<Self>) -> Response<Body> {
        let buf = state.metrics_manager.gather();

        Response::builder()
            .header(header::CONTENT_TYPE, "text/plain; version=0.0.4")
            .body(Body::from(buf))
            .unwrap()
    }
}

#[async_trait::async_trait]
impl HttpServer for AxumServer {
    async fn start(&mut self) -> Result<(), Error> {
        // If `host` converting to `Ipv4Addr` faild, then panic directly.
        let addr: Ipv4Addr = self.pisa_config.get_admin().host.parse().unwrap();
        let port = self.pisa_config.get_admin().port;
        let socket_addr = SocketAddr::new(IpAddr::V4(addr), port);
        info!("http server is binding port: {}", port);
        axum::Server::bind(&socket_addr)
            .serve(self.routes().into_make_service())
            .await
            .map_err(|e| ErrorKind::Runtime(e.into()))?;
        Ok(())
    }
}

File: pisa-controller/cmd/pisa-controller/http/http.go
// Copyright 2022 SphereEx Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package http

import (
	"flag"
	"fmt"
	"net/http"
	"time"

	"github.com/database-mesh/pisanix/pisa-controller/pkg/core"
)

var Config core.Config

func init() {
	flag.StringVar(&Config.Port, "corePort", "80", "CoreServer port.")
}

const (
	DefaultReadTimeout  = 5 * time.Second
	DefaultWriteTimeout = 10 * time.Second
)

type HttpConfig struct {
	Addr         string
	ReadTimeout  time.Duration
	WriteTimeout time.Duration
	TLSCertFile  string
	TLSKeyFile   string
}

func NewHttpConfig() *HttpConfig {
	return &HttpConfig{}
}

func (c *HttpConfig) SetAddr(port string) *HttpConfig {
	c.Addr = fmt.Sprintf(":%s", port)
	return c
}

func (c *HttpConfig) SetReadTimetout(t time.Duration) *HttpConfig {
	c.ReadTimeout = t
	return c
}

func (c *HttpConfig) SetWriteTimeout(t time.Duration) *HttpConfig {
	c.WriteTimeout = t
	return c
}

func (c *HttpConfig) SetTLSCertFile(file string) *HttpConfig {
	c.TLSCertFile = file
	return c
}

func (c *HttpConfig) SetTLSKeyFile(file string) *HttpConfig {
	c.TLSKeyFile = file
	return c
}

type HttpServer struct {
	core *http.Server
}

func (s *HttpServer) WithHandler(handler http.Handler) *HttpServer {
	s.core.Handler = handler
	return s
}

func (s *HttpServer) Build() *HttpServer {
	return s
}

func (s *HttpServer) ListenAndServe() error {
	return s.core.ListenAndServe()
}

func (s *HttpServer) ListenAndServeTLS(cert, key string) error {
	return s.core.ListenAndServeTLS(cert, key)
}

func NewHttpServer(conf *HttpConfig) *HttpServer {
	return &HttpServer{
		core: &http.Server{
			Addr:         conf.Addr,
			ReadTimeout:  DefaultReadTimeout,
			WriteTimeout: DefaultWriteTimeout,
		},
	}
}
